### ISSUE TITLE
experiment: compare base vs Synthetic Factory vs public SWE vs Prometheus vs mixtures under matched data budgets

### ISSUE THREAD

**Issue by rmems** [rmems/agoge-forger#103]
## Planning / research attribution

**GPT-5.6-Sol (OpenAI / ChatGPT), 2026-08-28**, with Raul Montoya Cardenas directing the data-source ablation and no-degradation research contract.

## Parent / existing infrastructure

Parent training program: #76 — first-class supervised fine-tuning baseline.

Reuse rather than duplicate:

* #99 — immutable train/validation/held-out split contract + leakage guards
* #100 — reproducible base-vs-SFT evaluation harness
* #101 — first controlled SFT pilot with frozen model/dataset contract

Synthetic producer program:

* `rmems/synthetic-factory#153`
* synthetic release candidate: `rmems/synthetic-factory#155`

Real engineering producer/eval program:

* `rmems/operation-prometheus#54` — issue-to-patch SFT materialization from real public GitHub engineering history
* `rmems/operation-prometheus#55` — time-forward, lineage-safe held-out engineering tasks

External public-data input:

* selected curated public SWE / coding-agent corpora from Hugging Face, admitted only after license, schema, provenance, leakage, and deduplication review

## Research question

> Does training on a curated oracle-grounded Synthetic Factory software-engineering corpus improve useful behavior without unacceptable degradation, and does it add incremental value beyond strong external public SWE corpora and provenance-rich real engineering trajectories from Operation Prometheus?

This issue evaluates the **data pipeline and data-source composition as interventions**. It is not a contest over which upstream generator model produced individual synthetic records.

The flagship question is:

> When a model is already trained on strong public SWE data plus curated real engineering history, does adding Synthetic Factory data improve transfer or at least preserve performance within a predeclared non-inferiority margin?

## Source taxonomy

Do not collapse every non-Synthetic-Factory dataset into a `real` bucket.

Use three provenance classes:

```text
SYNTHETIC_FACTORY
controlled generated tasks / trajectories
oracle-grounded and curated by rmems/synthetic-factory

EXTERNAL_PUBLIC_SWE
third-party curated SWE / coding-agent corpora
may itself contain synthetic agent trajectories over real or synthetic tasks

PROMETHEUS_REAL
rmems/operation-prometheus
real public GitHub engineering-history trajectories with lineage/provenance
```

The exact provenance class must survive into manifests and reports.

## Candidate external public SWE pool

The initial candidate pool includes, subject to source-specific admission review:

* `nvidia/Open-SWE-Traces`
* `togethercomputer/CoderForge-Preview`
* `open-thoughts/OpenThoughts-Agent-SFT-100K`
* `nvidia/SWE-Hero-openhands-trajectories`
* `SWE-bench/SWE-smith-trajectories`
* `nebius/SWE-rebench-openhands-trajectories`
* `R2E-Gym/R2EGym-SFT-Trajectories`
* `AmanPriyanshu/tool-reasoning-sft-CODING-MEnvData-SWE-Trajectory-data-cleaned-rectified`
* `voidful/agent-sft` only if its exact license/source mixture is approved
* `Manusagents/gpt-5.6-sol-coding-and-debugging-traces` only as an external model-generated/agentic source, never mislabeled as real engineering provenance

This list is a candidate pool, not automatic inclusion. Freeze the exact admitted source set and revisions before measured training.

Any additional dataset must pass the same admission contract rather than being concatenated ad hoc.

## External-source admission contract

Before an external dataset can enter the primary comparison, record and verify:

* exact Hugging Face repo + immutable revision;
* license and redistribution/training compatibility;
* schema and loader compatibility;
* task/repository/source provenance where available;
* whether records are real-history, synthetic-task, synthetic-trajectory, or mixed provenance;
* hidden/internal-reasoning policy and required stripping/transformation;
* exact duplicate and near-duplicate status;
* overlap with Operation Prometheus lineages;
* overlap with the held-out evaluation pool;
* source-specific quality/exclusion policy;
* normalization transform version and output digest.

No dataset enters because it is popular, large, or has a high download count.

## Required experimental arms

Using the exact same pinned starting checkpoint and frozen evaluation contract:

```text
A — BASE
no fine-tuning

B — SYNTHETIC FACTORY ONLY
Synthetic Factory #155 curated release

C — EXTERNAL PUBLIC SWE ONLY
approved third-party curated SWE / coding-agent corpus mixture

D — PROMETHEUS ONLY
Operation Prometheus #54 curated real-world SFT materialization

E — PUBLIC + PROMETHEUS
strong non-Synthetic-Factory baseline

F — SYNTHETIC + PUBLIC + PROMETHEUS
same total training budget as E for the primary composition test
```

## Primary causal comparisons

Report at least:

```text
B - A  does our synthetic pipeline help relative to no SFT?
C - A  what do established public SWE corpora buy?
D - A  what does provenance-rich real engineering history buy?
E - C  does Prometheus add value beyond public SWE?
F - E  does Synthetic Factory add value beyond the strongest public+real baseline?
```

**F - E is the flagship Synthetic Factory comparison.**

If F regresses relative to E beyond the predeclared margin, that is evidence against the current synthetic-data composition/curation policy and must be reported as such.

## Matched-budget rule

The primary B/C/D/E/F comparison must control training exposure as tightly as practical.

Freeze before measured results:

* accepted training-token budget per trained arm;
* optimizer and LR;
* effective update budget / epochs;
* sequence length;
* LoRA/QLoRA method and target modules;
* batch/gradient accumulation;
* seed policy;
* checkpoint selection;
* prompt/template version;
* decoding/evaluation settings.

If a source cannot provide enough accepted tokens to hit the common budget, do not silently oversample it. Record the shortfall and either lower the common primary budget or version a separate scaling experiment.

## Mixture contract

Freeze all first-pass mixture ratios before measured results are inspected.

A reasonable initial matched-budget design, if admitted-token availability permits:

```text
E — PUBLIC + PROMETHEUS
50% external public SWE tokens
50% Prometheus tokens

F — SYNTHETIC + PUBLIC + PROMETHEUS
~33% Synthetic Factory tokens
~33% external public SWE tokens
~33% Prometheus tokens
```

Both E and F use the same total accepted training-token budget in the primary comparison.

This means F-vs-E tests **data composition at fixed training exposure**, not the trivial effect of adding more tokens.

A later additive-scaling experiment may hold the E corpus fixed and append synthetic tokens, but that must be reported separately because additional token exposure is a confounder.

## Optional curation-ablation follow-up

After the primary A–F experiment, test whether Synthetic Factory's strict curation itself matters:

```text
F_loose  — public + Prometheus + minimally filtered synthetic data
F_strict — public + Prometheus + strict oracle-grounded curated synthetic data
```

Match accepted training-token budgets.

This asks whether the factory's validation/curation machinery adds value beyond merely increasing synthetic volume.

Do not run this follow-up until the primary A–F contract is complete.

## Contamination / held-out boundary

The primary engineering evaluation should consume frozen lineage-safe Prometheus held-out tasks from `operation-prometheus#55`, plus at least one protected regression/capability slice independent of all training sources.

No held-out lineage may:

* enter the Prometheus training arm;
* seed a Synthetic Factory repo/task pack;
* appear as a near-duplicate external public training task;
* appear as a near-duplicate synthetic training task;
* leak post-cutoff patches, reviews, checks, or outcomes into prompts.

Public datasets must be checked for overlap against held-out repositories/tasks before admission.

Use #99 split manifests and source-group metadata to fail closed.

## Cross-source normalization and deduplication

External public sources must not simply be concatenated.

Required flow:

```text
source-specific normalization
→ canonical source/provenance fields
→ exact duplicate removal
→ SWE lineage/provenance grouping where available
→ near-duplicate audit
→ quality/admission stratification
→ frozen source manifest
```

Keep deterministic provenance-based leakage guarantees separate from heuristic semantic near-duplicate checks.

## No-degradation contract

Before the measured A baseline is inspected, freeze:

* primary targeted engineering metric(s);
* protected regression metric(s);
* non-inferiority / degradation margin(s);
* invalid/inconclusive handling;
* evaluation sample membership;
* seed / replication policy;
* interpretation rules for mixed results.

Protected metrics must include at least one capability/regression slice not constructed from Synthetic Factory, Operation Prometheus, or the admitted external public SWE training sources so a narrow engineering gain cannot hide broad collapse.

Do not use training loss as proof of behavioral improvement.

### Pilot

One frozen run per trained arm may validate experiment plumbing and estimate variance.

### Confirmatory stage

Any claim that an intervention is improved/non-inferior should be replicated under a predeclared seed policy when compute permits. Near-margin outcomes remain inconclusive until replication or uncertainty analysis justifies interpretation.

Do not change the hypothesis, split, source membership, metric, or degradation margin after seeing pilot results to manufacture a positive result.

## Required measurements

For each arm preserve at minimum:

```text
exact model + revision
dataset releases + immutable revisions
source taxonomy + manifest hashes
accepted training tokens
record counts / mixture weights
normalization/dedup transform versions
training config + seed
training/validation loss where useful
held-out task success
base-to-arm delta
E-to-F delta
regressions / new successes
protected regression metrics
invalid/inconclusive eval count
wall-clock training time
tokens/sec
peak VRAM
checkpoint / adapter hashes
software/CUDA/GPU environment
```

## Outcome vocabulary

The report must be able to conclude:

```text
improved
non-inferior / preserved
mixed
null
regressed
inconclusive
```

Negative results are valid research artifacts.

## Required artifacts

Suggested structure:

```text
reports/synthetic-public-prometheus-ablation/<experiment_id>/
  experiment-contract.json
  source-admission-manifest.json
  source-manifests.json
  arm-A-base/
  arm-B-synthetic/
  arm-C-public/
  arm-D-prometheus/
  arm-E-public-prometheus/
  arm-F-all/
  comparisons.json
  regressions.jsonl
  contamination-report.json
  report.md
  reproduction.md
```

## Acceptance criteria

- [ ] Exact starting checkpoint/revision is identical across all trained arms.
- [ ] Exact external public source set and immutable revisions are frozen before measured training.
- [ ] External public datasets are provenance-classified rather than automatically called `real`.
- [ ] Every external source passes license/schema/provenance/leakage admission checks.
- [ ] B/C/D/E/F use the same matched accepted-token budget for the primary comparison, or a documented common lower budget if one source is smaller.
- [ ] E/F mixture ratios are frozen before results.
- [ ] Prometheus held-out lineages remain excluded from every training source, including public and synthetic derivations.
- [ ] Base and every trained arm use the identical held-out evaluation contract.
- [ ] Protected no-degradation metrics and margins are frozen before measured outcomes are inspected.
- [ ] Pairwise A–F deltas and regressions are machine-readable.
- [ ] F - E is reported explicitly as the incremental Synthetic Factory composition effect over the strongest public+real baseline.
- [ ] The report can honestly publish positive, negative, null, mixed, or inconclusive results.
- [ ] No Synthetic Factory expansion claim is made solely from training loss, dataset size, repository count, or factory-side quality scores.

## Relationships

* **Parent: **#76
* **Blocked by:** #99, #100, #101, `synthetic-factory#155`, `operation-prometheus#54`, `operation-prometheus#55`
* **Research parent: **`synthetic-factory#153`
* **Related roadmap: **#75
* **External public inputs:** admitted Hugging Face SWE/agent datasets under the source-admission contract above

## Tracking / Linear sync

GitHub is source of truth. Linear twin rmems/agoge-forger#103 tracks the same experiment in the existing Agoge Model Forge project under M7. Preserve cross-repo blockers and relationships in Linear rather than creating another twin.

## Ownership and dependency update (2026-08-30)

The external public-SWE scouting/admission work moved from closed #110 to `rmems/synthetic-factory#162`. PR #111 must transfer its evidence there rather than merge the research files into Agoge.

This experiment is now blocked by:

1. #102 — first published SFT intervention;
2. `rmems/synthetic-factory#155` — verified synthetic SWE release candidate;
3. `rmems/synthetic-factory#162` — pinned public-SWE admission manifest;
4. `rmems/operation-prometheus#54` — derived real-history SFT materialization;
5. `rmems/operation-prometheus#55` — lineage-safe held-out tasks.

Agoge owns only the frozen matched-budget arms, execution, evaluation, and report. Source discovery, generation, curation, and publication remain upstream-owner responsibilities.

**Comment by rmems** [comment]
## Public Dataset Scout decision — 2026-08-29

The 2026-08-21 scout has crossed from discovery into implementation-ready source admission.

`nebius/SWE-rebench-V2` is now the leading canonical `EXTERNAL_PUBLIC_SWE` source for this A–F experiment. The provisional license- and contamination-filtered core recipe is **22,042 rows**:

- Rust: 2,670
- Python: 5,495
- Go: 5,228
- TypeScript/JavaScript: 7,214
- Java: 1,435

Optional Julia/C/C++ rows remain spice, not primary arms.

Key negative findings remain important: `rustbench` is license-unknown with 145/500 Multi-SWE Rust-board overlap; LCA-CI remains eval-side; RepoDebug/OpenSRE/KernelBench/IaC-Eval/PolyBench remain evaluation candidates rather than `REAL_ONLY` training sources; no clean real CUDA, incident, Terraform-core, or human-review→repair arm was found.

Implementation is now tracked in **#110**. That child owns the bounded comparison of canonical V2 vs V2-PRs vs `PrimeIntellect/SWE-rebench-V2-Filtered-Verified`, fail-closed license admission, evaluation/Prometheus/first-party contamination exclusions, deterministic cross-source deduplication, and the frozen public-SWE source manifest. It must feed #99 rather than implement another splitter.

The three original scout artifacts are being preserved under `research/public-dataset-scout/2026-08-21/` and linked back to #110.

Research attribution: Grok Public Dataset Scout for discovery/census; GPT-5.6 Sol for #103/#110 integration and source-admission decomposition.

**Comment by rmems** [comment]
Reopening as a deliberately deferred research backlog item, not current execution work.

The issue remains blocked behind the first published evidence-grade SFT baseline and all upstream data contracts:

`#112 → #99 → #100 → #104 → #101 → #102 → #103`

Additional blockers: `rmems/synthetic-factory#155`, `rmems/synthetic-factory#162`, `rmems/operation-prometheus#54`, and `rmems/operation-prometheus#55`.

This corrects the mismatch between the retained roadmap and the prior `not planned` state. No ablation work should begin until every blocker is complete.

Planning/status correction: **OpenAI ChatGPT — Codex, GPT-5.6-Sol**, 2026-08-30, directed by **Raul Montoya Cardenas**.

**Comment by rmems** [comment]
**DELETE** — 04 Agoge Technical Debt Reducer

Debt-mode KEEP/DELETE triage for `rmems/agoge-forger#103`.
This is a training/experiment/roadmap item. Agoge maintenance mode forbids new training mills, model downloads, eval campaigns, and cloud jobs. Recommend close unless Raul reopens with a named train brief.

**Comment by rmems** [comment]
## Research note (freeze) — 2026-09-04

**KEEP** as closed matched-budget experiment design (planning only).

Does not claim constructible REAL_ONLY / REAL_SYNTHETIC_MIX arms. Locked: REAL_ONLY still **0/0**; V2 recipe-not-arm; SYNTHETIC_ONLY remains the only constructible snapshot arm; Prometheus neither-until-split; Multi-SWE eval-only undownloaded. No train / no parquet download from this note.
